scipy/signal/windows/_windows.py

```diff
@@ -170,16 +193,18 @@ def boxcar(M, sym=True):
     >>> plt.xlabel("Normalized frequency [cycles per sample]")
 
     """
+    xp = _namespace(xp)
+
     if _len_guards(M):
-        return np.ones(M)
+        return xp.ones(M, dtype=xp.float64)
     M, needs_trunc = _extend(M, sym)
 
-    w = np.ones(M, float)
+    w = xp.ones(M, dtype=xp.float64)
 
-    return _truncate(w, needs_trunc)
+    return xp.asarray(_truncate(w, needs_trunc), device=device)
```

Comment:
We need to take `device` into account when `_len_guards(M)`, and should also use that device for the whole calculation, rather than creating arrays on the CPU and potentially shifting devices at the end.

```diff
diff --git a/scipy/_lib/array_api_compat b/scipy/_lib/array_api_compat
index 498f086568..5affae52d6 160000
--- a/scipy/_lib/array_api_compat
+++ b/scipy/_lib/array_api_compat
@@ -1 +1 @@
-Subproject commit 498f086568362002185b005a0a7f38ad136ca8bb
+Subproject commit 5affae52d6095cb466394cd4f1b24678e952f578
diff --git a/scipy/signal/windows/_windows.py b/scipy/signal/windows/_windows.py
index 5120ee34c0..cc3ea97a93 100644
--- a/scipy/signal/windows/_windows.py
+++ b/scipy/signal/windows/_windows.py
@@ -196,12 +196,12 @@ def boxcar(M, sym=True, *, xp=None, device=None):
     xp = _namespace(xp)
 
     if _len_guards(M):
-        return xp.ones(M, dtype=xp.float64)
+        return xp.ones(M, dtype=xp.float64, device=device)
     M, needs_trunc = _extend(M, sym)
 
-    w = xp.ones(M, dtype=xp.float64)
+    w = xp.ones(M, dtype=xp.float64, device=device)
 
-    return xp.asarray(_truncate(w, needs_trunc), device=device)
+    return _truncate(w, needs_trunc)
```

Reply:
Yeah. Reworked in the last commit.